### ISSUE TITLE
:srgb or :icc-profile set causes black crush and strange gamma

### ISSUE THREAD

**Issue by Cyberbeing** [mpv-player/mpv#1563]
If this is for some reason intentional, it would be nice to have an option to enable gamut correction and linear light scaling, without forced gamma correction (i.e. input=output gamma).

[Normal Black/White Clipping] (http://i.imgbox.com/NBps4iL6.png)
[:sRGB Black/White Clipping] (http://i.imgbox.com/s1QzLFGd.png)

[Normal Color Steps] (http://i.imgbox.com/vK9wXpvd.png)
[:sRGB Color Steps] (http://i.imgbox.com/VB75gN83.png)

[:icc-profile Black/White Clipping] (http://i.imgbox.com/gdbig4me.png)
[madVR 3DLUT Black/White Clipping] (http://i.imgbox.com/nTx8JZhA.png)

[:icc-profile Color Steps] (http://i.imgbox.com/IcbEemZR.png)
[madVR 3DLUT Color Steps] (http://i.imgbox.com/vKi8B7fn.png)

The most noticeable issue with *:srgb* and *:icc-profile* on these AVS HD 709 test clips, is rather significant  near-black crush and less distinction between some of the steps. Trying to fix things with the *:gamma=* option just makes things worse.

**Comment by haasn** [comment]
What ICC profile were you testing it with?

**Comment by Cyberbeing** [comment]
I tried *:gamma* values between 0.9 & 1.1. When gamma is reduced below 1.0 with *:srgb* set, black tones become even more crushed with overall image too dark. When raised above 1.0 white tones quickly started to become crushed and overall image too bright. I was unable to normalize the gamma to the appearance of my *normal* images, since the curve being adjusted seems completely different.

The ICC profile was the same Argyll CMS BT.1886 calibrated LAB monitor profile which was used to create the madVR 3DLUT. Since *:srgb* without an ICC profile experiences similar black crush near-black, I assume mpv must be performing similar gamma handling with both *:srgb* & *:icc-profile=* to cause it. 

The primary reason I opened this issue was this black-crush/gamma issue with *:srgb* alone set, so I'd prefer to troubleshoot that first before involving another variable with an icc profile.

**Comment by haasn** [comment]
:icc-profile=whatever.icc:gamma=0.89136 should theoretically get you identical output to madVR, if using the same profile.

I'm not sure what's supposed to be wrong with the :srgb screenshots. Do you get a different output from other programs?

**Comment by haasn** [comment]
For reference, try the above command with this profile: http://www.nand.wakku.to/bt709/bt709-gamma2.2.icc

**Comment by wm4** [comment]
Isn't ``:srgb`` the flag that means "we do something no other media player does because we are super elitist"? So it's expected that the output is different. Anyway, I'm out again, I just wanted to display my ignorance.

**Comment by Cyberbeing** [comment]
> I'm not sure what's supposed to be wrong with the :srgb screenshots.

The near-black tones are crushed in the first set of images are crushed.

Without :srgb set, 19 step (R4-G4-B4), 18 step (R2-G2-B2), 17 step (R1-G1-B1), TV Black (R0-G0-B0)

With :srgb set, the 19 step (R1-G1-B1), 18 step (R0-G0-B0), 17 step (R0-G0-B0), TV Black (R0-G0-B0)

**Comment by haasn** [comment]
:srgb is functionally equivalent to :icc-profile with an sRGB profile. You could cross-compare to using :icc-profile=sRGB.icc, and ditto for madVR.

**Comment by haasn** [comment]
I did the math, and it seems to be correct:

```
λ let srgb_compand c | c <= 0.0031308 = 12.92*c | otherwise = (1+a) * c**(1/2.4) - a where a = 0.055
λ srgb_compand ((4/255) ** 1.961) * 255
0.9532724529793524
λ srgb_compand ((2/255) ** 1.961) * 255
0.24484837257782555
```

As you can clearly see, 4/4/4 compands to 1/1/1 under sRGB, and 2/2/2 compands to 0/0/0, which matches the results you get.

**Comment by Cyberbeing** [comment]
>:srgb is functionally equivalent to :icc-profile with an sRGB profile. You could cross-compare to using :icc-profile=sRGB.icc, and ditto for madVR.

I don't know why I would cross compare against madVR, when there is no standard behavior for creating a 3DLUT. If I set my monitor profile to sRGB and cross-compare against Photoshop, output looks identical to my *normal* mpv image. Not crushed like the *:srgb* MPV image.

>I did the math, and it seems to be correct

I very much doubt Photoshop's math is incorrect. So you must be doing something other than sRGB->Linear->sRGB. I assume you are mixing a BT.1886 function in there somewhere, or assuming the encoded gamma is different from the output gamma when converting to and from linear light. Which brings me back to why I opened this issue:

> If this is for some reason intentional, it would be nice to have an option to enable gamut correction and linear light scaling, without forced gamma correction (i.e. input=output gamma).

madVR for example offers linear light scaling and simple gamut correction without a 3DLut, which does not affect gamma. I'd like to see something like that supported in MPV.

**Comment by haasn** [comment]
Mpv supports that kind of logic internally, but why would we expose it to the user? If you have an ICC profile for your device, use an ICC profile. If you don't, and you care about accurate colors, get an ICC profile.

I don't see any point in introducing unnecessary extra abstractions. Might as well remove :srgb to eliminate issues like these, to be honest, since it's not like anything uses sRGB in practice either way (and if it did, you can use an ICC profile).

> So you must be doing something other than sRGB->Linear->sRGB. I assume you are mixing a BT.1886 function in there somewhere, or assuming the encoded gamma is different from the output gamma when converting to and from linear light.

I posted the calculation, why assume? We take the blu-ray (or whatever) source material and decompand it with a pure power curve (1.961) that is an approximation of the BT.709 OETF - aka BT.1886 without the gamma drop. (As for why we do this by default, there's a very lengthy discussion at #534)

Then you take this linear light result and compand it to the sRGB color space, resulting in the values I posted.

**Comment by Cyberbeing** [comment]
> Mpv supports that kind of logic internally, but why would we expose it to the user? If you have an ICC profile for your device, use an ICC profile.

Since not crushing blacks is desired behavior? If I encode a video with a particular color, I expect that exact color to be displayed on screen when I measure the display? I've already calibrated my display for BT.1886 gamma on the assumption of unmodified visually linear BT.709 video, so mpv modifying gamma a second time making output visually non-linear makes everything worse?

___
My basic desires are:

Option to workaround near-blacks or near-whites being crushed with :icc-profile or :srgb. Black point compensation?

Option for linear light scaling without requiring gamut or gamma correction.

Option to use :icc-profile gamut correction with without BT.1886 (or any other) secondary gamma correction performed by mpv.
___

@haasn if you are unwilling to make any of those changes, this issue can be closed.

**Comment by haasn** [comment]
mpv doesn't do “correction”, we do color management. The input (video) is mapped as accurately as possible onto the output (monitor), with as little distortion as possible.

Can you take a screenshot of your results when using this configuration? ```icc-profile:bt709-gamma2.4.icc:gamma=0.81708```

You can obtain the bt709-gamma2.4.icc profile here: http://www.nand.wakku.to/bt709-gamma2.4.icc

**Comment by Cyberbeing** [comment]
> Can you take a screenshot of your results when using this configuration? icc-profile:bt709-gamma2.4.icc:gamma=0.81708

[:icc-profile BT.709 2.4 Black/White Clipping] (http://i.imgbox.com/Z6tlRhwC.png)
[:icc-profile BT.709 2.4 Color Steps] (http://i.imgbox.com/GS9f3429.png).

**Comment by Cyberbeing** [comment]
>mpv doesn't do “correction”, we do color management.

Semantics... I'll modify my previous post to say "adaption" instead of "correction" if you feel that is more technically correct, but the intention of my feature requests are still the same. Color Management has two parts, "gamut adaption" and "gamma adaption" which can be performed independently of each other at different portions of the display chain (renderer, gpu, video post-processor, or display). Currently mpv is set up to perform color management entirely within the renderer, which is problematic if you are performing the required gamma adaption elsewhere. Additionally, mpv is making fixed assumptions about the encoded gamma of the video, which may not always hold true. 

For example, let's say I mastered a video on my calibrated BT.1886 (2.4) GDM-F520 CRT with gamut adaption from BT.709 primaries, under my intended viewing conditions. If I then encoded and immediately viewed said video, output with mpv :icc-profile would be incorrect and introduce unwanted distortions. While gamut adaption would be required for viewing, gamma adaptation would not, since my mastering monitor is already set at exactly the intended gamma for color managed viewing.

For another example, let's say I encoded a video where I desired a faint colored lights/noise moving across the background in pitch-black darkness. I assign these lights/noise dithered Rec709->RGB values between R0-G0-B0 and R2-G2-B2, which I verified would be clearly visible under pitch-black lighting conditions in Photoshop. Upon encoding my video and viewing with mpv color management, suddenly this entire scene would no longer be visible.

**Comment by haasn** [comment]
Comparing the results of the :icc-profile:gamma case with the “normal” screenshot, it doesn't seem like there's much of a difference anymore - the remaining few differences could likely be eliminated by picking a slightly different :gamma factor. (Although it seems like the tint changes slightly, that might be caused from the 3dlut-size settings. By default we don't cover the full 256x256x256 spectrum as precisely, since rounding errors are generally not that noticeable for really dark colors like this)

If anything, it doesn't appear there is a black crush anymore. Can you confirm?

> Color Management has two parts, "gamut adaption" and "gamma adaption"

This is not in line with the modern abstraction of color management, based on ICC profiles. Color management there has two distinct parts: 1. Source -> PCS and 2. PCS -> Display, each of which can be a black box as far as mpv is concerned. For PCS -> Display, we use the user-supplied ICC profile. Our only aim is to get the source material into the PCS.

We do not support any other model of color management (such as manually specifying gamut maps) that is not based on the ICC profile concept.

> For example, let's say I mastered a video on my calibrated BT.1886 (2.4) GDM-F520 CRT with software gamut adaption from BT.709 primaries, under my intended viewing conditions.

For your intended viewing conditions (ie. 10 Lux D65 ambient lighting), you should be specifying :gamma=81708. (See also the documentation of :gamma)

mpv's default behavior is to ignore system gamma, so if you want to replicate a BT.1886-based mastering workflow you should be using :gamma to introduce the system gamma (~1.22) again.

> Upon encoding my video and viewing with mpv color management, suddenly this entire scene would no longer be visible.

What, do you mean if you are using :srgb, or if you are using :icc-profile? For the former: stop using :srgb then, for the latter: see above

**Comment by haasn** [comment]
I took some screenshots (appropriately tagged):

mpv --pause 2-APL\ Clipping.mp4 -vo opengl-hq: ![](http://i.srsfckn.biz/37b.png)
mpv --pause 2-APL\ Clipping.mp4 -vo opengl-hq:icc-profile=bt709-gamma2.4.icc:gamma=0.81708333:3dlut-size=256x256x256:dither-depth=8: ![](http://i.srsfckn.biz/3c.png)

I don't think there's a discernible black crush present. Some very mild brightness drop, probably due to a rounding error of some sort. What do you think?

**Comment by Cyberbeing** [comment]
>icc-profile=bt709-gamma2.4.icc:gamma=0.81708333:3dlut-size=256x256x256:dither-depth=8

It still darkens near-blacks by around 1R-1G-1B on average, but it's close. 

Though with 3-Color Steps.mp4 it seems to be slightly changing hue of the primary & secondary color steps. turning colors which should be 0 into values between 6-12, which seems a bit problematic.

**Comment by haasn** [comment]
Not sure where the left-over differences are coming from. A lot of error is probably introduced by the fact that we generated the internal 3DLUT from a much wider color space than the source (BT.2020, to be precise), and rely on LittleCMS to perceptually desaturate colors to fit into the new gamut. I'm not sure how reliably that preserves the saturation of colors which are very close to the edge of the gamut.

It might be worth doing some experiments to see if it still introduces this discoloration when using a much wider color space as target.

But just to be clear, though: Which parts of the image in particular are you referring to?

**Comment by Cyberbeing** [comment]
> But just to be clear, though: Which parts of the image in particular are you referring to?

All of the primary and secondary color patches seem to show this [hue change] (http://i.imgbox.com/wyWC8EE7.png) with 
\Misc Patterns\A - Additional\3-Color Steps.mp4
and the previous command line. The neutral tones seem unaffected.

**Comment by haasn** [comment]
What do you mean when you say “primary and secondary color patches”?

Also, earlier you requested the ability to do gamut adaption without gamma adaption. What's your use case for this? And furthermore, please realize that this is inherently incompatible with ICC profiles, since ICC profiles are a black box. We can't just apply “half” of the profile. What would your recommended API for this be? Specifying all 8 color coordinates on the command line?

> Option for linear light scaling without requiring gamut or gamma adaption (color management).

Upon reconsideration, this might actually be a worthwhile feature to implement.

**Comment by haasn** [comment]
I've decoupled linear light scaling from CMS in https://github.com/mpv-player/mpv/commit/4fed18e81e7ffca623e4260d8569f959410b10bb since I agree that they are reasonably separate features.

It is now enabled by default in opengl-hq mode (since it's implied by sigmoid-upscaling).

**Comment by Cyberbeing** [comment]
> What do you mean when you say “primary and secondary color patches”?

The color steps. If you take a color sampler to that image, and measure the various colors.

*Red* color steps should be only contain red, yet they all have a minor amounts of green and blue.
*Green* color steps should be only contain green, yet they all have a minor amounts of red and blue.
*Blue* color steps should be only contain blue, yet they all have a minor amounts of red and green..
*Magenta* color steps should be only contain red and blue, yet they all have a minor amounts of green.
*Yellow* color steps should be only contain red and green, yet they all have a minor amounts of blue.
*Cyan* color steps should be only contain green and blue, yet they all have a minor amounts of red.

This occurs only when *icc-profile* is enabled in mpv, even though the video is BT.709 and the icc profile is BT.709 which both should have essentially identical gamuts and not require an adaption of primary and secondary colors like this.

>Also, earlier you requested the ability to do gamut adaption without gamma adaption. What's your use case for this? And furthermore, please realize that this is inherently incompatible with ICC profiles, since ICC profiles are a black box. We can't just apply “half” of the profile. What would your recommended API for this be? Specifying all 8 color coordinates on the command line?

Something simple and shader based like madVR's "this display is already calibrated" setting, for when the user is aware of general gamut of the display but does not have means to calibrate the display and create an icc profile. In these cases, simple gamut adaption, even if just clipping can be useful.

madVR for example, then allows the user to set the gamut of their display via simple preset (BT.709, SMTPE-C, EBU/PAL, BT.2020, DCI-PC), as well a gamma (Pure power curve or True BT.709 curve with a scaled gamma range between 1.8 - 2.6). When the properties of the video being played do not match the user settings, a simple adaption is performed in shaders. While you could do something more advanced which allow specifying all 8 color coordinates, I think simple presets such as madVR would better serve this purpose.

While you could potentially do something similar within the ICC color managed environment, since you would essentially need to modify the gamma curve of the input profiles on the fly to match the gamma of the users output monitor profile. Alternatively, you could always use a fixed gamma user selected gamma for the input profile, and override the output monitor profile to specify that same gamma. This may or may not be practical to do, but since mpv users lcms which is capable of generating profiles, I assume it would be somehow possible. If it's too complex within lcms, you could do it outside standard color management with simple shaders as above.

>I've decoupled linear light scaling from CMS in 4fed18e since I agree that they are reasonably separate features.

Thank you. Next request related to this, would be allowing to specify when linear light scaling is used. Something like *:linear-scaling= upscale|downscale|both* since as least personally, I'd prefer to use only linear downscaling but not upscaling.

**Comment by haasn** [comment]
I've wished for a way to specify monitor profile “presets” in the past, and some recent changes to the way ICC profiles are handled would actually make this possible.

Regardless, I don't think we should be seeing this much error in the current icc-profile based approach. I'd rather stick to one abstraction (ICC profiles). It's possible that there are two optimizations we could make, from separate ends:

1. Allow generating the 3DLUT against a different PCS than BT.2020 (if this is the cause of the issue)
2. Allow specifying the monitor profile as a “preset” or “virtual profile” (eg. BT.709 primaries, gamma 2.35 or whatever)

I don't really want to “short circuit” the LittleCMS codepath at all, as we currently do with :srgb. With the plan I have, we could get rid of :srgb and essentially just make it one of many possible ICC profile presets.

To be fair, though, a good argument *against* this approach is performance. Even under the assumption that we can fine-tune the 3DLUT approach to get its output visually identical to “in shader” color management, it would still probably be slower. I haven't really measured by how much, though. Probably nowhere near the complexity of upscaling, but still nontrivial.

**Comment by Cyberbeing** [comment]
>Allow specifying the monitor profile as a “preset” or “virtual profile” (eg. BT.709 primaries, gamma 2.35 or whatever)

The only issue I have with solutions such as this, is they are usually limiting in terms of ability to describe custom monitor gamma curves to the extent that it would bypass gamma adaption. If you are sticking with ICC profiles, we would need some kind of tool which can add fake gamma information to monitor profiles which will match what mpv uses as the gamma of the source profile. 

Alternatively, you would need a way to actually extract accurate profile gamma data points (not calibration curves) from matrix & 3dlut ICC profiles which do not utilize a power-curve, and provide it in a format which mpv can understand. Realistically the resulting data is unlikely to be compact enough to provide via cli, and would likely require loading of some kind of external 'gamma' text file for mpv to utilize as the inverse encoding gamma. I'm just brainstorming here though.

>To be fair, though, a good argument against this approach is performance. 

We would have to see, but considering the extremely low gpu usage currently required I doubt it would be anything crippling to modern GPUs.

> I could rework the logic to make :linear-scaling only downscale in linear light in the absence of :sigmoid-upscaling, but I feel that goes a bit against intuition.

I just remember from experience that performing linear light upscaling brings out the worse aspects of pretty much any upscaling algorithm, causing issues such as extreme halos/ringing and aliasing around edges and detail which really put me off from using linear light it for anything but downscaling with low-ringing algorithms.

I've yet to investigate if :sigmoid-upscaling resolves enough general fault & artifact issues with using normal linear light upscaling though to make using it worthwhile over gamma light. The technical benefits of linear light alone would not justify it, if it still produces noticeable artifacts. Unless you can say with confidence that :sigmoid-upscaling linear upscaling does not in any way increase overall ringing & aliasing compared to gamma light scaling with all supported upscaling algorithms, I would see a :linear-downscaling option as having its usefulness.

**Comment by haasn** [comment]
For what it's worth, I think it should be possible via the LittleCMS API to open up a profile and modify it after it was loaded (eg. by changing the gamma characteristics), as long as it's not an overcomplicated LUT profile or something. But I don't think that's something we want to be doing either way.

I will most likely not be adding any logic to load external gamma tables, 3dluts or gamut descriptions from text files or whatever, I still think that's too much unwarranted complexity where it's not needed. I do not see the practical limitations of the ICCv4 profile format that doing gamut adaption manually would solve.

> We would have to see, but considering the extremely low gpu usage currently required I doubt it would be anything crippling to modern GPUs.

I'm guessing you haven't played around with any of the advanced upscaling options. :)

> I've yet to investigate if :sigmoid-upscaling resolves enough general fault & artifact issues with using normal linear light upscaling though to make using it worthwhile over gamma light.

In my experience, it actually makes them (being artifacts like ringing) *less bad* than normal gamma-based upscaling. Our logic is mostly based on what ImageMagick recommends for upscaling algorithms.

**Comment by Cyberbeing** [comment]
> I'm guessing you haven't played around with any of the advanced upscaling options. :)

I have, but it still uses only ~9-10% GPU load at non-boost clocks scaling from 1280x720->2560x1440.
opengl-hq
ewa_lanczos
scale-antiring
sigmoid-upscaling
fbo-format=rgba16f
icc-profile
3dlut-size=256x256x256
dither-depth=8

Remove icc-profile & 3dlut-size, and GPU load is ~8% a difference of only ~1-2% which is negligible.

> In my experience, it actually makes them (being artifacts like ringing) less bad than normal gamma-based upscaling. Our logic is mostly based on what ImageMagick recommends for upscaling algorithms.

Okay, I'll check this out later. Could you explain the purpose and effects of the sigmoid-center & sigmoid-slope values? The manual is extremely vague about the relationship of the sigmoid curve to the behavior of :sigmoid-upscaling,

**Comment by haasn** [comment]
Heh, I guess your GPU must be more powerful than mine. I have a HD7950 and ewa_lanczos for scale+cscale (both at radius 3) is just barely as much as it can handle, at radius 4 I start dropping frames.

> Could you explain the purpose and effects of the sigmoid-center & sigmoid-slope values?

![](https://upload.wikimedia.org/wikipedia/commons/thumb/8/88/Logistic-curve.svg/320px-Logistic-curve.svg.png)

This is what the sigmoid curve looks like. sigmoid-slope determines the slope (dy/dx) at the center (y=0.5), and sigmoid-center determines where on the x axis said center is located.

It's hard to quantify the effect of this, but there's a trade-off of some sort. Lower slope values are less effective at preventing artifacts, but higher slope values introduce a negative version of the artifacts. You can experiment with the values somewhat, higher slopes can get you a better result on some content, but the defaults we picked are what we determined to be the “safest”.

**Comment by wm4** [comment]
Ping?

**Comment by haasn** [comment]
I want to implement the second half of this after all, perhaps after the gl_rework stuff.

I've thought about the API, and I think it should be something like “display-colorspace=<value>” where value is the number corresponding to one of various colorspaces we have implemented (it would be a direct mapping of enum MP_CSP_PRIMARIES).

If specified, this would determine the PCS used for the 3DLUT generation as well - in essence, icc-profile would simply be another way to fine-tune the colors on top of this selection.

If left disabled/default, it would use BT.2020 for the 3DLUT, and do nothing otherwise.

Thoughts?

**Comment by haasn** [comment]
I've implemented this in gl_rework: https://github.com/mpv-player/mpv/commit/62efba5e6b7e0215e2415bc2397e021e5e5f4300

The API is to let you use ```target-prim=<value>``` and ```target-trc=<value>``` to specify the characteristics of your display, in a way completely distinct from (and mutually exclusive with) icc-profile. (You don't have to use both)

I intend on also using target-prim for controlling the icc-profile source, but this is not implemented yet. (Nor is icc-profile implemented at all in gl_rework)

Instead of using magic numbers, I've opted for a saner list of choices instead. Assuming this lands in master, does it solve this issue?
